### DISCUSSION TITLE
Avoiding processing on alpha channel

### DISCUSSION THREAD

**Discussion by yuguang-li** [halide/Halide#1388] [General]
I'm writing high-performance image processing filters in Halide. The input and output buffers are all defined in rgb-alpha four channels. When I'm processing the input image, the generated C code  loops from channel 0 to channel 3. This happens in every single function in my code. I did tried to use `out.bound(0,3)` in the processing step and use `final(x,y,c) = select(c<3,out(x,y,c),255)` in the end. But the code cannot be compiled correctly. Does anyone have the same issue when doing optimization?

**Comment by abadams** [comment]
For simple pointwise filters, the number of channels that get evaluated is
determined by the number of channels in your output image. If you set
output.extent[2] == 3, it should only evaluate things over three channels.

In general Halide determines what to evaluate using the size of the output.

**Comment by yuguang-li** [comment]
Thanks for your response Adam. 
That's true. But what if my input has four channels, but I only want all my functions to work on RGB three channels? I'll add the alpha channel at the last step of my whole pipeline.

**Comment by abadams** [comment]
It doesn't matter how many channels the input has. The computation is
determined by how many channels the output has. Perhaps this is the pattern
you want:

output(x, y, c) = select(c < 3, input(x, y, clamp(c, 0, 2)), 255);

The clamp lets Halide know that you only need channels 0, 1, and 2 of all
previous pipeline stages.
